cloud/ec2_util.go

```diff
@@ -329,6 +314,85 @@ func cacheHostData(ctx context.Context, env evergreen.Environment, h *host.Host,
 	return nil
 }
 
+// cacheAllHostData is the same as cacheHostData but optimized for updating many
+// hosts at once.
+func cacheAllHostData(ctx context.Context, env evergreen.Environment, client AWSClient, pairs []hostInstancePair) error {
+	catcher := grip.NewBasicCatcher()
+
+	hostsToCache := map[string]host.CloudProviderData{}
+	for _, hostAndInstance := range pairs {
+		h := hostAndInstance.host
+		instance := hostAndInstance.instance
+		data, err := makeCloudProviderData(instance)
+		if err != nil {
+			catcher.Wrapf(err, "making cloud provider data for host '%s'", h.Id)
+			continue
+		}
+
+		hostsToCache[h.Id] = *data
+
+		h.Zone = data.Zone
+		h.StartTime = data.StartedAt
+		h.Host = data.PublicDNS
+		h.PublicIPv4 = data.PublicIPv4
+		h.IPv4 = data.PrivateIPv4
+		h.IP = data.IPv6
+		h.Volumes = data.Volumes
+
+		if h.NoExpiration {
+			// This is not a bulk operation for convenience because it's assumed
+			// that the number of unexpirable hosts is small.
+			grip.Error(message.WrapError(setHostPersistentDNSName(ctx, env, h, utility.FromStringPtr(instance.PublicIpAddress), client), message.Fields{
+				"message":    "could not update host's persistent DNS name",
+				"op":         "upsert",
+				"dashboard":  "evergreen sleep schedule health",
+				"host_id":    h.Id,
+				"started_by": h.StartedBy,
+			}))
+		}
+	}
+
+	catcher.Wrap(host.CacheAllCloudProviderData(ctx, env, hostsToCache), "bulk caching host data")
+
+	return catcher.Resolve()
+}
+
+func makeCloudProviderData(instance *types.Instance) (*host.CloudProviderData, error) {
+	if err := validateCloudProviderData(instance); err != nil {
+		return nil, err
```

Comment:
Can we wrap this with a message saying validation failed?

Reply:
Done.